lib/async-iterator.js

```diff
@@ -0,0 +1,444 @@
+  }
+
+  async forEach(fn, thisArg) {
+    for await (const value of this) {
+      await fn.call(thisArg, value);
+    }
+  }
+
+  async every(predicate, thisArg) {
+    for await (const value of this) {
+      if (!await predicate.call(thisArg, value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  async find(predicate, thisArg) {
+    for await (const value of this) {
+      if (await predicate.call(thisArg, value)) {
+        return value;
+      }
+    }
+    return undefined;
+  }
+
+  async includes(element) {
+    for await (const value of this) {
+      if (value === element || (isNaN(value) && isNaN(element))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  async reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = await this.next();
+      if (next.done) {
+        throw new TypeError(
+          'Reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for await (const value of this) {
+      result = await reducer(result, value);
+    }
+    return result;
+  }
+
+  async some(predicate, thisArg) {
+    for await (const value of this) {
+      if (await predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  async someCount(predicate, count, thisArg) {
+    let n = 0;
+    for await (const value of this) {
+      if (await predicate.call(thisArg, value)) {
+        if (++n === count) return true;
+      }
+    }
+    return false;
+  }
+
+  async collectTo(CollectionClass) {
+    const arr = await this.toArray();
+    return new CollectionClass(arr);
+  }
+
+  async collectWith(obj, collector) {
+    await this.forEach(element => collector(obj, element));
+  }
+
+  async join(sep = ',', prefix = '', suffix = '') {
+    let result = prefix;
+    const { done, value } = await this.next();
+    if (!done) {
+      result += value;
+      for await (const value of this) {
+        result += sep + value;
+      }
+    }
+    return result + suffix;
+  }
+
+  async toArray() {
+    const newArray = [];
+    for await (const value of this) {
+      newArray.push(value);
+    }
+    return newArray;
+  }
+
+  map(mapper, thisArg) {
+    return new MapIterator(this, mapper, thisArg);
+  }
+
+  filter(predicate, thisArg) {
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
+    return new FlatMapIterator(this, mapper, thisArg);
+  }
+
+  zip(...iterators) {
+    return new ZipIterator(this, iterators);
+  }
+
+  chain(...iterators) {
+    return new ChainIterator(this, iterators);
+  }
+
+  take(amount) {
+    return new TakeIterator(this, amount);
+  }
+
+  takeWhile(predicate, thisArg) {
+    return new TakeWhileIterator(this, predicate, thisArg);
+  }
+
+  skip(amount) {
+    return new SkipIterator(this, amount);
```

Comment:
This can be implemented as
```js
for (let i = 0; i < amount; i++) {
  this.next();
}
return this;
```
This will both simplify the implementation and avoid the creation of a `SkipIterator` class instance.